torch/_library/abstract_impl.py

```diff
@@ -0,0 +1,197 @@
+import contextlib
+import functools
+import warnings
+from typing import Callable, Optional
+
+import torch
+from torch._library.utils import Kernel, RegistrationHandle
+
+
+class AbstractImplHolder:
+    """A holder where one can register an abstract impl to."""
+
+    def __init__(self, qualname: str):
+        self.qualname: str = qualname
+        self.kernel: Optional[Kernel] = None
+        self.lib: Optional[torch.library.Library] = None
+
+    def register(self, func: Callable, source: str) -> RegistrationHandle:
+        """Register an abstract impl.
+
+        Returns a RegistrationHandle that one can use to de-register this
+        abstract impl.
+        """
+        if self.kernel is not None:
+            raise RuntimeError(
+                f"impl_abstract(...): the operator {self.qualname} "
+                f"already has an abstract impl registered at "
+                f"{self.kernel.source}."
+            )
+        if torch._C._dispatch_has_kernel_for_dispatch_key(self.qualname, "Meta"):
+            raise RuntimeError(
+                f"impl_abstract(...): the operator {self.qualname} "
+                f"already has an DispatchKey::Meta implementation via a "
+                f"pre-existing torch.library or TORCH_LIBRARY registration. "
+                f"Please either remove that registration or don't call "
+                f"impl_abstract."
```

Comment:
Hmm, you might legitimately want to override an incomplete meta with a fake tensor impl that can handle data dependent dynamic shapes. Well, we can deal with it if it ever happens.